File: analytics/src/main/java/com/bluetriangle/analytics/BtCrashHandler.java
package com.bluetriangle.analytics;

import android.os.Build;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.io.Writer;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;


final class BtCrashHandler implements Thread.UncaughtExceptionHandler {
    private final Thread.UncaughtExceptionHandler defaultUEH;

    private final BlueTriangleConfiguration configuration;

    private Timer crashHitsTimer;

    public BtCrashHandler(@NonNull final BlueTriangleConfiguration configuration) {
        this.defaultUEH = Thread.getDefaultUncaughtExceptionHandler();
        this.configuration = configuration;
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    @Override
    public void uncaughtException(@NonNull Thread t, Throwable e) {
        final String timeStamp = String.valueOf(System.currentTimeMillis());
        this.crashHitsTimer = new Timer().start();
        final String stacktrace = Utils.exceptionToStacktrace(null, e);

        try {
            sendToServer(stacktrace, timeStamp);
        } catch (InterruptedException interruptedException) {
            interruptedException.printStackTrace();
        }

        defaultUEH.uncaughtException(t, e);
    }

    private void sendToServer(final String stacktrace, final String timeStamp) throws InterruptedException {
        Thread thread = new Thread(new CrashRunnable(configuration, stacktrace, timeStamp, crashHitsTimer));
        thread.start();
        thread.join();
    }


}

File: analytics/src/main/java/com/bluetriangle/analytics/BtCrashHandler.kt
package com.bluetriangle.analytics

import android.os.Build
import androidx.annotation.RequiresApi
import com.bluetriangle.analytics.Utils.exceptionToStacktrace
import com.bluetriangle.analytics.deviceinfo.IDeviceInfoProvider

internal class BtCrashHandler(
    private val configuration: BlueTriangleConfiguration,
    private val deviceInfoProvider: IDeviceInfoProvider
) : Thread.UncaughtExceptionHandler {
    internal val defaultUEH: Thread.UncaughtExceptionHandler = Thread.getDefaultUncaughtExceptionHandler()

    @RequiresApi(api = Build.VERSION_CODES.N)
    override fun uncaughtException(t: Thread, e: Throwable) {
        val timeStamp = System.currentTimeMillis().toString()
        val mostRecentTimer = Tracker.instance?.getMostRecentTimer()
        configuration.logger?.debug("Most Recent Timer: $mostRecentTimer")

        val stacktrace = exceptionToStacktrace(null, e)
        try {
            sendToServer(mostRecentTimer, stacktrace, timeStamp)
        } catch (interruptedException: InterruptedException) {
            interruptedException.printStackTrace()
        }
        defaultUEH.uncaughtException(t, e)
    }

    @Throws(InterruptedException::class)
    private fun sendToServer(mostRecentTimer:Timer?, stacktrace: String, timeStamp: String) {
        val thread = Thread(CrashRunnable(
            configuration, stacktrace, timeStamp,
            mostRecentTimer = mostRecentTimer,
            deviceInfoProvider = deviceInfoProvider
        ))
        thread.start()
        thread.join()
    }
}